examples/error-handling/src/main.rs

```diff
@@ -0,0 +1,232 @@
+                .on_failure(()),
+        )
+        .with_state(state);
+
+    let listener = tokio::net::TcpListener::bind("127.0.0.1:3000")
+        .await
+        .unwrap();
+    tracing::debug!("listening on {}", listener.local_addr().unwrap());
+    axum::serve(listener, app).await.unwrap();
+}
+
+#[derive(Default, Clone)]
+struct AppState {
+    next_id: Arc<AtomicU64>,
+    users: Arc<Mutex<HashMap<u64, User>>>,
+}
+
+#[derive(Deserialize)]
+struct UserParams {
+    name: String,
+}
+
+#[derive(Serialize, Clone)]
+struct User {
+    id: u64,
+    name: String,
+    created_at: Timestamp,
+}
+
+async fn users_create(
+    State(state): State<AppState>,
+    // Make sure to use our own JSON extractor so we get input errors formatted in a way that
+    // matches our application
+    AppJson(params): AppJson<UserParams>,
+) -> Result<AppJson<User>, AppError> {
+    let id = state.next_id.fetch_add(1, Ordering::SeqCst);
+
+    // We have implemented `From<time_library::Error> for AppError` which allows us to use `?` to
+    // automatically convert the error
+    let created_at = Timestamp::now()?;
+
+    let user = User {
+        id,
+        name: params.name,
+        created_at,
+    };
+
+    state.users.lock().unwrap().insert(id, user.clone());
+
+    Ok(AppJson(user))
+}
+
+// Create our own JSON extractor by wrapping `axum::Json`. This makes it easy to override the
+// rejection and provide our own which formats errors to match our application.
+//
+// `axum::Json` responds with plain text if the input is invalid.
+#[derive(FromRequest)]
+#[from_request(via(axum::Json), rejection(AppError))]
+struct AppJson<T>(T);
+
+impl<T> IntoResponse for AppJson<T>
+where
+    axum::Json<T>: IntoResponse,
+{
+    fn into_response(self) -> Response {
+        axum::Json(self.0).into_response()
+    }
+}
+
+// The kinds of errors we can hit in our application.
+enum AppError {
+    // The request body contained invalid JSON
+    JsonRejection(JsonRejection),
+    // Some error from a third party library we're using
+    TimeError(time_library::Error),
+}
+
+// Tell axum how `AppError` should be converted into a response.
+//
+// This is also a convenient place to log errors.
+impl IntoResponse for AppError {
+    fn into_response(self) -> Response {
+        // How we want errors responses to be serialized
+        #[derive(Serialize)]
+        struct ErrorResponse {
+            message: String,
+        }
+
+        let (status, message) = match self {
+            AppError::JsonRejection(rejection) => {
+                // This error is caused by bad user input so don't log it
+                (rejection.status(), rejection.body_text())
+            }
+            AppError::TimeError(err) => {
+                // Because `TraceLayer` wraps each request in a span that contains the request
+                // method, uri, etc we don't need to include those details here
+                tracing::error!(%err, "error from time_library");
```

Comment:
Maybe controversial, but I find the `err=` bit in the output a bit weird. IMHO tracing fields are much more useful for fields whose exact value you'd want to match when searching logs, rather than error messages that are generally just searched for specific keywords.

```suggestion
                tracing::error!("error from time_library: {err}");
```